Media: can't be synced - missing meta table

```
java.lang.RuntimeException: An error occurred while executing doInBackground()
at android.os.AsyncTask$3.done(AsyncTask.java:354)
at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
at java.util.concurrent.FutureTask.run(FutureTask.java:271)
at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
at java.lang.Thread.run(Thread.java:764)
Caused by: android.database.sqlite.SQLiteException: no such table: meta (code 1): , while compiling: update meta set dirMod = ?
at io.requery.android.database.sqlite.SQLiteConnection.nativePrepareStatement(Native Method)
at io.requery.android.database.sqlite.SQLiteConnection.acquirePreparedStatement(SQLiteConnection.java:3)
at io.requery.android.database.sqlite.SQLiteConnection.prepare(SQLiteConnection.java:2)
at io.requery.android.database.sqlite.SQLiteSession.prepare(SQLiteSession.java:3)
at io.requery.android.database.sqlite.SQLiteProgram.<init>(SQLiteProgram.java:8)
at io.requery.android.database.sqlite.SQLiteStatement.<init>(SQLiteStatement.java:1)
at io.requery.android.database.sqlite.SQLiteDatabase.executeSql(SQLiteDatabase.java:2)
at io.requery.android.database.sqlite.SQLiteDatabase.execSQL(SQLiteDatabase.java:2)
at com.ichi2.utils.DatabaseChangeDecorator.execSQL(DatabaseChangeDecorator.java:3)
at com.ichi2.libanki.DB.execute(DB.java:6)
at com.ichi2.libanki.Media._logChanges(Media.java:12)
at com.ichi2.libanki.Media.findChanges(Media.java:3)
at com.ichi2.async.CollectionTask.doInBackgroundCheckMedia(CollectionTask.java:3)
at com.ichi2.async.CollectionTask.actualDoInBackground(CollectionTask.java:27)
at com.ichi2.async.CollectionTask.doInBackground(CollectionTask.java:2)
at com.ichi2.async.CollectionTask.doInBackground(CollectionTask.java:1)
at android.os.AsyncTask$2.call(AsyncTask.java:333)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
... 4 more
```

--- a/AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java
+++ b/AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java
@@ -1640,6 +1640,12 @@ public class CollectionTask extends BaseAsyncTask<TaskData, TaskData, TaskData> 
     private TaskData doInBackgroundCheckMedia() {
         Timber.d("doInBackgroundCheckMedia");
         Collection col = getCol();
+        // Ensure that the DB is valid - unknown why, but some users were missing the meta table.
+        try {
+            col.getMedia().rebuildIfInvalid();
+        } catch (IOException e) {
+            return new TaskData(false);
+        }
         // A media check on AnkiDroid will also update the media db
         col.getMedia().findChanges(true);
         // Then do the actual check

--- a/AnkiDroid/src/main/java/com/ichi2/libanki/Media.java
+++ b/AnkiDroid/src/main/java/com/ichi2/libanki/Media.java
@@ -24,9 +24,11 @@ import android.text.TextUtils;
 
 import android.util.Pair;
 
+import com.ichi2.anki.AnkiDroidApp;
 import com.ichi2.libanki.template.Template;
 import com.ichi2.utils.Assert;
 
+import com.ichi2.utils.ExceptionUtil;
 import com.ichi2.utils.JSONArray;
 import com.ichi2.utils.JSONObject;
 
@@ -1018,4 +1020,35 @@ public class Media {
         long mod = mDb.queryLongScalar("select dirMod from meta");
         return mod == 0;
     }
+
+
+    public void rebuildIfInvalid() throws IOException {
+        try {
+            _changed();
+            return;
+        } catch (Exception e) {
+            if (!ExceptionUtil.containsMessage(e, "no such table: meta")) {
+                throw e;
+            }
+            AnkiDroidApp.sendExceptionReport(e, "media::rebuildIfInvalid");
+
+            // TODO: We don't know the root cause of the missing meta table
+            Timber.w(e, "Error accessing media database. Rebuilding");
+            // continue below
+        }
+
+
+        // Delete and recreate the file
+        mDb.getDatabase().close();
+
+        String path = mDb.getPath();
+        Timber.i("Deleted %s", path);
+
+        new File(path).delete();
+
+        mDb = new DB(path);
+        _initDB();
+    }
+
+
 }

--- a/AnkiDroid/src/main/java/com/ichi2/utils/ExceptionUtil.java
+++ b/AnkiDroid/src/main/java/com/ichi2/utils/ExceptionUtil.java
@@ -1,0 +1,32 @@
+/*
+ *  Copyright (c) 2020 David Allison <davidallisongithub@gmail.com>
+ *
+ *  This program is free software; you can redistribute it and/or modify it under
+ *  the terms of the GNU General Public License as published by the Free Software
+ *  Foundation; either version 3 of the License, or (at your option) any later
+ *  version.
+ *
+ *  This program is distributed in the hope that it will be useful, but WITHOUT ANY
+ *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+ *  PARTICULAR PURPOSE. See the GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along with
+ *  this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.ichi2.utils;
+
+public class ExceptionUtil {
+    public static boolean containsMessage(Throwable e, String needle) {
+        if (e == null) {
+            return false;
+        }
+
+        if (containsMessage(e.getCause(), needle)) {
+            return true;
+        }
+
+        String message = e.getMessage();
+        return message != null && message.contains(needle);
+    }
+}

--- a/AnkiDroid/src/test/java/com/ichi2/libanki/CheckMediaTest.java
+++ b/AnkiDroid/src/test/java/com/ichi2/libanki/CheckMediaTest.java
@@ -1,0 +1,54 @@
+/*
+ *  Copyright (c) 2020 David Allison <davidallisongithub@gmail.com>
+ *
+ *  This program is free software; you can redistribute it and/or modify it under
+ *  the terms of the GNU General Public License as published by the Free Software
+ *  Foundation; either version 3 of the License, or (at your option) any later
+ *  version.
+ *
+ *  This program is distributed in the hope that it will be useful, but WITHOUT ANY
+ *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+ *  PARTICULAR PURPOSE. See the GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along with
+ *  this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.ichi2.libanki;
+
+import com.ichi2.anki.RobolectricTest;
+import com.ichi2.async.CollectionTask;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import java.util.concurrent.ExecutionException;
+
+import androidx.test.ext.junit.runners.AndroidJUnit4;
+
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+@RunWith(AndroidJUnit4.class)
+public class CheckMediaTest extends RobolectricTest {
+
+    @Override
+    protected boolean useInMemoryDatabase() {
+        return false;
+    }
+
+
+    @Test
+    public void checkMediaWorksAfterMissingMetaTable() throws ExecutionException, InterruptedException {
+        // 7421
+        getCol().getMedia().getDb().getDatabase().execSQL("drop table meta");
+
+        assertThat(getCol().getMedia().getDb().queryScalar("SELECT count(*) FROM sqlite_master WHERE type='table' AND name='meta';"), is(0));
+
+        CollectionTask task = CollectionTask.launchCollectionTask(CollectionTask.TASK_TYPE.CHECK_MEDIA);
+
+        task.get();
+
+        assertThat(getCol().getMedia().getDb().queryScalar("SELECT count(*) FROM sqlite_master WHERE type='table' AND name='meta';"), is(1));
+    }
+}

